src/univariate/discrete/geometric.jl

```diff
@@ -34,7 +34,9 @@ function Geometric(p::Real; check_args::Bool=true)
     return Geometric{typeof(p)}(p)
 end
 
-Geometric() = Geometric{Float64}(0.5)
+struct OneHalf <: Real end
+Geometric() = Geometric{OneHalf}(OneHalf())
```

Comment:
There's some runtime cost there but if you'd prefer runtime branching over compile time branching I can switch the implementation. Here's an idea of the runtime cost:

```julia-repl
julia> using Random, Chairmarks

julia> current(p) = floor(Int,-randexp() / log1p(-p))
current (generic function with 1 method)

julia> pr() = leading_zeros(rand(UInt))
pr (generic function with 1 method)

julia> proposed(p) = p == 1//2 ? pr() : current(p)
proposed (generic function with 1 method)

julia> @b .2 current,proposed
(8.764 ns, 9.627 ns)

julia> @b .5 current,proposed
(8.582 ns, 2.584 ns)

julia> @b pr
2.280 ns
```

Reply:
I think the runtime cost is fine (overhead of ~0.3 ns it seems?). It's still a significant improvement and it would be consistent with existing samplers and `rand` implementations in Distributions which would also be a bit simpler for users I assume.

Reply:
Done. The runtime comparison was pretty substantial in bulk generation but inlining allows the compiler to hoist it, giving another 3x speedup:

```julia-repl
julia> using Distributions

julia> @b Geometric() rand(_, 100)
891.906 ns (2 allocs: 928 bytes)

julia> @eval Distributions function rand(rng::AbstractRNG, d::Geometric)
           if d.p == 0.5
               leading_zeros(rand(rng, UInt)) # This branch is a performance optimization
           else
               floor(Int,-randexp(rng) / log1p(-d.p))
           end
       end
rand (generic function with 181 methods)

julia> @b Geometric() rand(_, 100)
342.896 ns (2 allocs: 928 bytes)

julia> @eval Distributions @inline function rand(rng::AbstractRNG, d::Geometric)
           if d.p == 0.5
               leading_zeros(rand(rng, UInt)) # This branch is a performance optimization
           else
               floor(Int,-randexp(rng) / log1p(-d.p))
           end
       end
rand (generic function with 181 methods)

julia> @b Geometric() rand(_, 100)
110.390 ns (2 allocs: 928 bytes)

(@v1.11) pkg> st Distributions
Status `~/.julia/environments/v1.11/Project.toml`
  [31c24e10] Distributions v0.25.115
```